tensorzero_internal/src/cache.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::HashMap;
+
+use crate::clickhouse::ClickHouseConnectionInfo;
+use crate::error::{Error, ErrorDetails};
+use crate::inference::types::batch::deserialize_json_string;
+use crate::inference::types::{ContentBlock, ModelInferenceRequest, ModelInferenceResponse};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct CacheOptions {
+    #[serde(default)]
+    pub enabled: bool,
+    #[serde(default)]
+    pub lookback_s: Option<u32>,
+}
+
+#[derive(Debug, Clone)]
+pub struct ModelProviderRequest<'request> {
+    pub request: &'request ModelInferenceRequest<'request>,
+    pub model_name: &'request str,
+    pub provider_name: &'request str,
+}
+
+impl ModelProviderRequest<'_> {
+    pub fn get_cache_key(&self) -> Result<[u8; 32], Error> {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(self.model_name.as_bytes());
+        hasher.update(self.provider_name.as_bytes());
+        let serialized_request = serde_json::to_string(self.request).map_err(|e| {
+            Error::new(ErrorDetails::Serialization {
+                message: format!("Failed to serialize request: {e}"),
+            })
+        })?;
+        let request_bytes = serialized_request.as_bytes();
+        hasher.update(request_bytes);
+        Ok(*hasher.finalize().as_bytes())
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct ModelInferenceCacheRow {
+    short_cache_key: u64,
+    long_cache_key: String,
+    output: Vec<ContentBlock>,
+    raw_request: String,
+    raw_response: String,
+}
+
+// This doesn't block
+pub fn start_cache_write(
+    clickhouse_client: &ClickHouseConnectionInfo,
+    request: ModelProviderRequest<'_>,
+    output: &[ContentBlock],
+    raw_request: &str,
+    raw_response: &str,
+) -> Result<(), Error> {
+    let cache_key = request.get_cache_key()?;
+    let short_cache_key = u64::from_le_bytes(cache_key[..8].try_into().map_err(|e| {
```

Comment:
What do you think about making `get_cache_key` return a `CacheKey` type, which has a `get_long` and `get_short` method? I'm worried about the `[..8]` in `start_cache_write` and `cache_lookup` getting out of sync

Reply:
I think that is good and will add